includes/routes/class-order-post.php

```diff
@@ -0,0 +1,264 @@
+	}
+
+	/**
+	 * Create the order.
+	 *
+	 * @param WP_REST_Request $request JSON request for shipping endpoint.
+	 *
+	 * @return array|WP_Error|WP_REST_Response
+	 */
+	protected function create_order( $request ) {
+		$wc_rest_orders_controller = new \WC_REST_Orders_Controller();
+		$order_response            = $wc_rest_orders_controller->create_item( $request );
+
+		return \is_wp_error( $order_response ) ? $order_response : $order_response->data;
+	}
+
+	/**
+	 * Set the shipping line item on the order.
+	 *
+	 * @param WC_Data         $order   Object object.
+	 * @param WP_REST_Request $request Request object.
+	 *
+	 * @return WC_Data
+	 */
+	public function set_order_shipping( $order, $request ) {
+		// Do nothing if there are no shipping options or shipping rates.
+		if ( empty( $this->shipping_options ) || empty( $this->shipping_options['shipping_rates'] ) ) {
+			return $order;
+		}
+
+		$shipping_rates = $this->shipping_options['shipping_rates'];
+		$shipping_rate  = $this->get_best_shipping_rate( $shipping_rates );
+
+		if ( ! empty( $shipping_rate ) ) {
+
+			// Clear the existing shipping line items.
+			$items = $order->get_items( 'shipping' );
+
+			if ( count( $items ) > 0 ) {
+				// Loop through shipping items.
+				foreach ( $items as $item_id => $item ) {
+					$order->remove_item( $item_id );
+				}
+				$order->calculate_totals();
+			}
+
+			// Set the array for tax calculations.
+			$calculate_tax_for = array(
+				'country'  => $order->get_shipping_country(),
+				'state'    => $order->get_shipping_state(),
+				'postcode' => $order->get_shipping_postcode(),
+				'city'     => $order->get_shipping_city(),
+			);
+
+			// Set a total shipping amount.
+			$new_ship_price = floatval( $shipping_rate['price'] );
+
+			// Get a new instance of the WC_Order_Item_Shipping Object.
+			$item = new \WC_Order_Item_Shipping();
+
+			$item->set_method_title( $shipping_rate['name'] );
+			$item->set_method_id( $shipping_rate['rate_id'] );
+			$item->set_total( $new_ship_price );
+			$item->calculate_taxes( $calculate_tax_for );
+
+			// Add the new shipping item to the order.
+			$order->add_item( $item );
+
+			// Recalculate the totals.
+			$order->calculate_totals();
+		}
+
+		return $order;
+	}
+
+	/**
+	 * Get the best rate from the list of shipping rates.
+	 *
+	 * @param array $shipping_rates The list of shipping rates.
+	 *
+	 * @return array
+	 */
+	protected function get_best_shipping_rate( $shipping_rates ) {
+		if ( empty( $shipping_rates ) ) {
+			return array();
+		}
+
+		$best_rate            = PHP_INT_MAX;
+		$best_rate_index      = false;
+		$shipping_rates_count = count( $shipping_rates );
+
+		if ( 1 === $shipping_rates_count ) {
+			$best_rate_index = 0;
+		}
+
+		if ( false === $best_rate_index ) {
+			foreach ( $shipping_rates as $index => $shipping_rate ) {
+				// Do not choose the local pickup rate.
+				if ( 'local_pickup' === $shipping_rate['method_id'] ) {
+					continue;
```

Comment:
If local pickup is the only option we should select it. does this best_rate_index logic cover that scenario?

Reply:
I'll have to check. I'm pretty sure I just copied this from the `shipping` endpoint, so if the logic is wrong it will need to be changed there as well.

Reply:
👍

Reply:
Ok, this is new code based on [this Notion doc](https://www.notion.so/fastaf/WC-single-outbound-request-checkout-proposal-a5f13f285e0447bd9e7863ab4490caed#c100a36138ec4eecae4fac1e0321e01f).

Just reviewing the logic, and if local pickup is the only option, we do select it. That's what lines 196-198 do for us. I think we're good to go with this logic. @adibalwani has tested this endpoint before and confirmed that it worked.

Reply:
Sounds good